### ISSUE TITLE
Redesign uiautomator Java parsing

### ISSUE THREAD

**Issue by bootstraponline** [appium/appium#4858]
I think we'd be better off with a [json schema](http://json-schema.org/) approach to parsing custom uiautomator commands. In uiautomator2, they've added a bunch of new classes including [BySelector](http://developer.android.com/reference/android/support/test/uiautomator/BySelector.html). If we continue to add custom parsing code, it's going to be a mess. Java is difficult to parse and doesn't have an easy way of being dynamically evaluated on Android.

Another issue with the current parsing code is it's inflexible. I can't express complicated queries such as look for a resourceId on API 18 however don't crash if we're looking for a resourceId on a lower API level because I've also used a fallback selector. This is an actual bug. #316

Regular UiSelector parsing has also regressed and/or had some bugs.
- https://github.com/appium/appium/issues/4311
- https://github.com/appium/appium/issues/4482

I think if we were simply parsing JSON then the complexity of this could be dramatically reduced. The architecture would also be more flexible (although the commands would be different) when moving to different backends (uiautomator2, droiddriver). @Jonahss what do you think?

**Comment by Jonahss** [comment]
Well it would certainly make the code and execution easier on the Java side. Creating the equivalent of a Java `eval()` function (which is what we've done, in a limited scope) is cool, but pretty costly and a little bit crazy.

On the other hand, I'm willing to do crazy things for our users because I'm worried about usability. It's hard enough to understand the UiSelector system itself, let alone a JSON schema built on top of it. Need I remind you that the old 'complex find' method which was logical and concise while being completely unintelligible to pretty much everybody.

If we put in the effort to create a schema which was very well thought out, did not have strange corner-cases, and was elegant, I would be for it. But we would have to put great great effort into properly documenting it and explaining to all of our users how it works.

**Comment by 0x1mason** [comment]
Custom query language compiled at runtime a la https://code.google.com/p/image-playground/

**Comment by bootstraponline** [comment]
> Need I remind you of the old 'complex find' method which was logical and concise while being completely unintelligible to pretty much everybody?

I agree it was unintelligible however it worked. The current system is an improvement. Parsing strings is easier to understand however it's buggy, hard to extend, and makes certain use cases impossible. I'm not sure that's the right trade off.

Ideally evaling code (similar to how execute script works on iOS) would be possible.

> If we put in the effort to create a schema which was very well thought out, did not have strange corner-cases, and was elegant, I would be for it. But we would have to put great great effort into properly documenting it and explaining to all of our users how it works.

I agree it'd require extraordinary effort. I'm hoping someone has a better idea.

**Comment by bootstraponline** [comment]
[calabash invented a custom query language](https://github.com/calabash/calabash-android/blob/327220e6b00512e918a9dc1fd2ded7579b2fc4e5/ruby-gem/test-server/instrumentation-backend/src/sh/calaba/instrumentationbackend/query/antlr/UIQueryParser.java) powered by ANTLR that works cross platform. One option is to do something similar. Probably less crazy than parsing strings of Java code.

**Comment by 0x1mason** [comment]
Here are some ideas, some might be terrible, but throwing them out for sake
of discussion:

1) Don't expose Android APIs at all.  Instead, allow users to right sql
queries against xml.
2) They have xpath.  They don't need another way to query.
3) Use the Java ScriptingEngine and [expose relevant Java classes and
objects via JS.](http://www.ibm.com/developerworks/library/j-5things9/)
Could also use [Rhino](
https://developer.mozilla.org/en-US/docs/Mozilla/Projects/Rhino/Overview).
Here's an example of [Rhino on Android](
https://vec.io/posts/embed-javascript-in-android-java-code-with-rhino).
4) Use the Java ScriptingEngine and [expose relevant Java classes and
objects via Scala.](
https://issues.scala-lang.org/browse/SI-874) Or Jython, JRuby, etc.
5) Require users to supply a precompiled jar with interface X.  The jar
contains their queries.  Their tests pass string references that we use to
invoke the queries by reflection.

My inclination is 3 or 5. 4 would be fun, but using scala might be too
complicated for some users.  5 is the easiest and perhaps the most powerful
option, but requires users know Java.



On Sun, Apr 5, 2015 at 9:54 PM, bootstraponline <notifications@github.com>
wrote:

> calabash invented a custom query language
> <https://github.com/calabash/calabash-android/blob/327220e6b00512e918a9dc1fd2ded7579b2fc4e5/ruby-gem/test-server/instrumentation-backend/src/sh/calaba/instrumentationbackend/query/antlr/UIQueryParser.java>
> powered by ANTLR that works cross platform. One option is to do something
> similar. Probably less crazy than parsing strings of Java code.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/appium/appium/issues/4858#issuecomment-89887595>.
>



-- 

*Eric Millin*
*Software Engineer, Ecosystems*
Twitter: @etmillin <http://twitter.com/jlipps>
GitHub: @0x1mason <https://github.com/jlipps>
Sauce Labs <http://saucelabs.com/> | Sauce blog <http://sauceio.com/>

**Comment by 0x1mason** [comment]
Here are some ideas, some might be terrible, but throwing them out for sake of discussion: 

1) Don't expose Android APIs at all. Instead, allow users to right sql queries against xml. 
2) They have xpath. They don't need another way to query.
3) Use the Java ScriptingEngine and [expose relevant Java classes and objects via JS.](http://www.ibm.com/developerworks/library/j-5things9/) Could also use [Rhino]( https://developer.mozilla.org/en-US/docs/Mozilla/Projects/Rhino/Overview). Here's an example of [Rhino on Android]( https://vec.io/posts/embed-javascript-in-android-java-code-with-rhino). 
4) Use the Java ScriptingEngine and [expose relevant Java classes and objects via Scala.]( https://issues.scala-lang.org/browse/SI-874) Or Jython, JRuby, etc.
5) Require users to supply a precompiled jar with interface X. The jar contains their queries. Their tests pass string references that we use to invoke the queries by reflection. 

My inclination is 3 or 5. 4 would be fun, but using scala might be too complicated for some users. 5 is the easiest and perhaps the most powerful option, but requires users know Java.

**Comment by Jonahss** [comment]
The string-parsing approach has one or two issues logged against it right now, they're easy to fix and we could prioritize those. I can also refactor the code to make it easier to add more methods. Just mentioning that the task isn't impossibly hard, but I am open to better solutions.

The Calabash query system looks very useful, but doesn't replace the original reasons we added the `uiselector` locator strategy. The point is to have a way of accessing specific methods in Android UiAutomator that don't fit within normal querying and to have a locator strategy that doesn't require learning something new for Android devs already familiar with UiAutomator.

Thanks for outlining a bunch of potential solutions @0x1mason. Here's my take:
1) XPath *is* a query language for indexing xml.

2) I'm all for just using XPath, but the one thing it can't do is Scroll. I say we add better scroll functions to the clients, then users can use XPath.

3) This would require people to learn something new, on top of UiSelector (which is already a questionable framework). But if people are interested in it we could go this way.

4) I think I would prefer 3 over this. But the whole point of Appium is to be *language agnostic* so requiring any specific language for a query is kinda odd.

5) This is beyond the scope of my Java knowledge, so I leave it up to experienced users currently running tests. I predict that the test writing and debugging workflow would be more difficult, but if devs are willing to accept that....

Ok so in review:

I think users should use XPath, and if something can't be done in XPath, we should add more utility methods to the clients (like scrolling). Does @bootstraponline have more examples of the sorts of things which can only be done with the UiSelector syntax?

If experienced Java users would prefer the Java ScriptingEngine exposing JS, or supplying precompiled jars which adhere to some interface, we could work on that route.

Do we want to take this to the Appium discuss forum? I feel like most of our strong Java users are there.

**Comment by 0x1mason** [comment]
Maybe all we need is a scrolling find method. If that's it, then adding
support for that would be the simplest route.

On Mon, Apr 6, 2015 at 1:25 PM, Jonah <notifications@github.com> wrote:

> The string-parsing approach has one or two issues logged against it right
> now, they're easy to fix and we could prioritize those. I can also refactor
> the code to make it easier to add more methods. Just mentioning that the
> task isn't impossibly hard, but I am open to better solutions.
>
> The Calabash query system looks very useful, but doesn't replace the
> original reasons we added the uiselector locator strategy. The point is
> to have a way of accessing specific methods in Android UiAutomator that
> don't fit within normal querying and to have a locator strategy that
> doesn't require learning something new for Android devs already familiar
> with UiAutomator.
>
> Thanks for outlining a bunch of potential solutions @0x1mason
> <https://github.com/0x1mason>. Here's my take:
> 1) XPath *is* a query language for indexing xml.
>
> 2) I'm all for just using XPath, but the one thing it can't do is Scroll.
> I say we add better scroll functions to the clients, then users can use
> XPath.
>
> 3) This would require people to learn something new, on top of UiSelector
> (which is already a questionable framework). But if people are interested
> in it we could go this way.
>
> 4) I think I would prefer 3 over this. But the whole point of Appium is to
> be *language agnostic* so requiring any specific language for a query is
> kinda odd.
>
> 5) This is beyond the scope of my Java knowledge, so I leave it up to
> experienced users currently running tests. I predict that the test writing
> and debugging workflow would be more difficult, but if devs are willing to
> accept that....
>
> Ok so in review:
>
> I think users should use XPath, and if something can't be done in XPath,
> we should add more utility methods to the clients (like scrolling). Does
> @bootstraponline <https://github.com/bootstraponline> have more examples
> of the sorts of things which can only be done with the UiSelector syntax?
>
> If experienced Java users would prefer the Java ScriptingEngine exposing
> JS, or supplying precompiled jars which adhere to some interface, we could
> work on that route.
>
> Do we want to take this to the Appium discuss forum? I feel like most of
> our strong Java users are there.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/appium/appium/issues/4858#issuecomment-90162514>.
>



-- 

*Eric Millin*
*Software Engineer, Ecosystems*
Twitter: @etmillin <http://twitter.com/jlipps>
GitHub: @0x1mason <https://github.com/jlipps>
Sauce Labs <http://saucelabs.com/> | Sauce blog <http://sauceio.com/>

**Comment by jlipps** [comment]
:+1: to simply adding a scrolling find method. It could simply be a new locator strategy (`-id with scroll`), that the clients adopt. We could either add support for this in iOS or make it android-only. The client code would look like:

```
el = driver.find_and_scroll_by_id('foo')
```

That way the semantics are transparent (this is a find with a side-effect). I think it's a kind of dumb way of architecting things in UiAutomator, but it's what we've got to work with.

**Comment by bootstraponline** [comment]
> The string-parsing approach has one or two issues logged against it right now, they're easy to fix and we could prioritize those.

I think it'd be good to fix those since that's what people are using today.

> Do we want to take this to the Appium discuss forum? I feel like most of our strong Java users are there.

We should pick an approach before asking for opinions. I think currently it's an open question on the best path forward.

> Maybe all we need is a scrolling find method.

I think we do need a scrolling find method. I have one in the Ruby bindings (that was ported to the Java bindings) which represents what I think is the common case for scrolling on Android. 

> Does @bootstraponline have more examples of the sorts of things which can only be done with the UiSelector syntax?

The reason for this issue is to discuss how we can expose the underlying native Android automation APIs in a way that's easy for people to use and for us to maintain. The scope is beyond finding or scrolling. I think [NativeScript](https://github.com/NativeScript/NativeScript) with their [0 day support for new Android APIs](https://www.nativescript.org/blog/answering-nativescript-beta-webinar-questions) combined with the fact they're using appium is one potential option. I'm going to check with them to see if they are willing to support the embedded use case. Plugging in NativeScript to the existing bootstrap jar would allow people to use any Android API via standard JavaScript which is the ideal. It's open source under Apache 2.0 as well.

Some users want to use only the [underlying automation methods](https://github.com/songkick/rubium-ios). The linked example is for iOS. This isn't going to be possible on Android with out a redesign to our approach. There are a number of use cases such as [UiWatchers](http://developer.android.com/tools/testing-support-library/index.html), [BySelectors](http://developer.android.com/reference/android/support/test/uiautomator/BySelector.html), [UiObject2Conditions](http://developer.android.com/reference/android/support/test/uiautomator/UiObject2Condition.html), which just don't fit into Selenium and probably never will. It'd enable people to explore different features and then bring back the useful bits into the appium server where they could be made more accessible. With the current approach, huge pieces of functionality are just not architecturally possible for people to even attempt using. I can't imagine what parsing a UiWatcher would look like.

**Comment by bootstraponline** [comment]
> I think users should use XPath

uiautomator (at least the version we're using) is unable to dump the ui hierarchy reliably. When uiautomator fails to do this, it'll crash the app. We have some hacks to try and prevent this however it doesn't cover all cases. Also uiautomator fails to create valid xml consistently (again we have a hack for this also) and returns incorrect elements when using xpath (no solution for this, although we've tried to make it better by using instance numbers instead of node indexes). In summary XPath is currently a pile of hacks upon hacks that barely works for the common cases and always fails down under advanced usage. It's also the slowest method to find elements. On iOS, the situation is worse where you can't even find elements such as buttons. **No one should be using XPath** until the backend is changed to something that is sane (uiautomator2 or droiddriver on android).

**Comment by bootstraponline** [comment]
I posted on the [NativeScript google group](https://groups.google.com/forum/#!topic/nativescript/kVjS867WOUI). If they don't support embedding then it might not even be an option.

**Comment by jlipps** [comment]
All good points @bootstraponline. We have a balance to strike here between presenting a coherent model to users and making sure they can avail themselves of whatever's useful. Right now, I'm in favor of (1) fixing the known bugs with the string parsing, and (2) adding a find+scroll method to server/clients. Who knows, this might get us quite a long way, while we work on migrating to UiAutomator2/DroidDriver; maybe those things will end up taking care of the longer term issues. We need to think _really_ carefully before breaking the WebDriver model as a workaround instead of as an intentional and consistent direction that we can talk clearly about.

**Comment by bootstraponline** [comment]
For better or worse, execute_script to run arbitrary JavaScript is clearly within the WebDriver model and already implemented on iOS. If anything, replacing Java parsing with JavaScript increases the coherence for users.

I agree with the priorities. The redesign is more of a long term issue and the path forward isn't clear yet. I think it's worth exploring options at the conceptual level even if we're not planning on implementing it in the short term.

It's possible that UiAutomator2 and DroidDriver are vastly more stable and don't suffer from the bugs we've been dealing with in the old UiAutomator. I don't think we'll know for sure until we've switched. The problems with old UiAutomator took a long time to discover and remedy.

**Comment by bootstraponline** [comment]
> Yes, it is possible and actually that's the way NativeScript for Android works. With a bit of simplification, integrating NativeScript for Android is as easy as calling the following two lines
>
>System.loadLibrary("NativeScript");
com.tns.Platform.init(context);
>
>An even easier option is to configure AndroidManifest.xml to use NativeScriptApplication and NativeScriptActivity (take a look at https://github.com/NativeScript/android-runtime/tree/master/test-app) though this option may not be the best one when more precise control flow is needed.
>
> I think NativeScript/Appium scenario will be of great use for everyone and I am willing to help with a proof-of-concept. I'll have some time after 2-3 weeks and will take a look at it. Meanwhile, I'd glad to answer any further questions.
> 
> https://groups.google.com/d/msg/nativescript/kVjS867WOUI/qU5-uBIoGswJ

I think this is worth exploring.

**Comment by Jonahss** [comment]
Looks good. Thanks for pointing out how being able to access the underlying Automation Framework is useful for other projects to try out new ideas, then integrate the successful ones more fully into appium. You're right that parsing strings is only going to get us so far, especially as we enter more sophisticated class structures within Android.

**Comment by bootstraponline** [comment]
@slavchev published an initial proof of concept. I think once 1.5 is ready then this is worth looking into.

- [appium bootstrap changes](https://github.com/slavchev/appium/commit/fcb2e2a4e37ff2c15dbc6bbfef2378169188754d#diff-8ff5cb905b00561d7c8ac966529fb854R24)
- [how to build](https://github.com/slavchev/android-runtime/blob/slavchev/appium-integration/docs/appium-integration.md)

**Comment by SrinivasanTarget** [comment]
@bootstraponline Can we close this?

**Comment by bootstraponline** [comment]
I think so. We replaced the Android backend.
